File: src/modules/audio/frame-buffer.ts
/**
 * Circular frame buffer keyed by timestamp.
 * Stores the last N milliseconds of frame pairs.
 * When audio trigger fires, flushes the ±windowMs window around the trigger.
 */

import type { FramePair } from '@/types/tracking';

export interface FrameBufferOptions {
  /** Total rolling buffer duration in ms (default 2000ms = 2s) */
  bufferDurationMs?: number;
  /** Window to flush around trigger: total = 2 * halfWindowMs (default ±500ms) */
  halfWindowMs?: number;
}

export interface BufferedFrame {
  pair: FramePair;
  wallTimeMs: number;
}

export type FlushCallback = (frames: FramePair[], contactWallTimeMs: number) => void;

export class FrameBuffer {
  private readonly ring: BufferedFrame[] = [];
  private readonly bufferDurationMs: number;
  private readonly halfWindowMs: number;
  private onFlush: FlushCallback | null = null;
  private pendingTriggerTime: number | null = null;

  constructor(opts: FrameBufferOptions = {}) {
    this.bufferDurationMs = opts.bufferDurationMs ?? 2000;
    this.halfWindowMs = opts.halfWindowMs ?? 500;
  }

  setFlushCallback(cb: FlushCallback): void {
    this.onFlush = cb;
  }

  /**
   * Push an incoming frame pair into the ring buffer.
   * Old frames outside the buffer window are evicted.
   */
  push(pair: FramePair): void {
    const wallTimeMs = performance.now();
    this.ring.push({ pair, wallTimeMs });

    // Evict frames older than buffer duration
    const cutoff = wallTimeMs - this.bufferDurationMs;
    while (this.ring.length > 0 && this.ring[0].wallTimeMs < cutoff) {
      this.ring.shift();
    }

    // If we have a pending trigger and enough post-trigger frames, flush
    if (this.pendingTriggerTime !== null) {
      const elapsed = wallTimeMs - this.pendingTriggerTime;
      if (elapsed >= this.halfWindowMs) {
        this._flush(this.pendingTriggerTime);
        this.pendingTriggerTime = null;
      }
    }
  }

  /**
   * Signal that audio trigger fired at this wall-clock time.
   * The flush will happen once sufficient post-trigger frames accumulate.
   */
  trigger(triggerWallTimeMs: number): void {
    this.pendingTriggerTime = triggerWallTimeMs;
  }

  private _flush(triggerTime: number): void {
    const start = triggerTime - this.halfWindowMs;
    const end   = triggerTime + this.halfWindowMs;

    const window = this.ring
      .filter((f) => f.wallTimeMs >= start && f.wallTimeMs <= end)
      .map((f) => f.pair);

    this.onFlush?.(window, triggerTime);
  }

  clear(): void {
    this.ring.length = 0;
    this.pendingTriggerTime = null;
  }
}

File: nuc/frame_buffer.py
"""
Rolling frame-pair buffer with audio-triggered flush.

Stores the last BUFFER_DURATION_S of stereo frame pairs.
When trigger() is called, waits for HALF_WINDOW_S of post-trigger
frames then calls the flush callback with the ±window.
"""

import threading
import time
from collections import deque
from typing import Callable, List, Optional

import config
from capture import FramePair

FlushCallback = Callable[[List[FramePair], float], None]


class FrameBuffer:
    def __init__(self, on_flush: FlushCallback) -> None:
        self._on_flush = on_flush
        self._ring: deque = deque()
        self._lock = threading.Lock()
        self._trigger_time: Optional[float] = None

    def push(self, pair: FramePair) -> None:
        now = time.monotonic()
        with self._lock:
            self._ring.append((now, pair))
            cutoff = now - config.BUFFER_DURATION_S
            while self._ring and self._ring[0][0] < cutoff:
                self._ring.popleft()

            if self._trigger_time is not None:
                elapsed = now - self._trigger_time
                if elapsed >= config.HALF_WINDOW_S:
                    self._flush_locked()

    def trigger(self, trigger_time: float) -> None:
        with self._lock:
            self._trigger_time = trigger_time

    def _flush_locked(self) -> None:
        t = self._trigger_time
        assert t is not None
        start = t - config.HALF_WINDOW_S
        end   = t + config.HALF_WINDOW_S

        window = [pair for (ts, pair) in self._ring if start <= ts <= end]
        self._trigger_time = None

        if window:
            # Hand off without holding the lock
            threading.Thread(
                target=self._on_flush,
                args=(window, t),
                daemon=True,
            ).start()

    def clear(self) -> None:
        with self._lock:
            self._ring.clear()
            self._trigger_time = None
